File: src/cc_wrap.py
import asyncio
import json
import logging
import os
from typing import List, Dict, Any, AsyncIterator, Tuple, Optional

logger = logging.getLogger(__name__)

# Default timeout for Claude Code invocations (in seconds)
# Can be overridden via timeout parameter
DEFAULT_TIMEOUT = 600  # 10 minutes

# Tools that raymond-managed agents must never use (they are orchestrator-level concerns)
DISALLOWED_TOOLS = ["EnterPlanMode", "ExitPlanMode", "AskUserQuestion", "NotebookEdit"]


class ClaudeCodeTimeoutError(Exception):
    """Raised when Claude Code invocation times out."""
    pass


def _build_claude_env() -> dict[str, str]:
    """Build environment for Claude Code subprocesses.

    Strips the CLAUDECODE env var so that claude doesn't think it's being
    nested inside another Claude Code session (raymond intentionally launches
    claude as a managed subprocess, not as a nested session).
    """
    env = os.environ.copy()
    env.pop("CLAUDECODE", None)
    return env


def _build_claude_command(
    prompt: str,
    model: Optional[str] = None,
    session_id: Optional[str] = None,
    dangerously_skip_permissions: bool = False,
    **kwargs
) -> List[str]:
    """Build the claude CLI command with all arguments.
    
    Args:
        prompt: The prompt to send to claude
        model: The model to use (e.g., "haiku", "sonnet", "opus")
        session_id: Optional session ID to resume an existing session
        dangerously_skip_permissions: If True, passes --dangerously-skip-permissions
            instead of --permission-mode acceptEdits. WARNING: This allows Claude
            to execute any action without prompting for permission.
        **kwargs: Additional arguments to pass to claude command.
            Supported kwargs include:
            - fork (bool): If True, passes --fork-session flag to branch from session_id
            - Any other kwargs are converted to --key value CLI arguments
        
    Returns:
        List of command arguments
    """
    cmd = [
        "claude",
        "-p",  # headless/print mode
        "--output-format", "stream-json",
        "--verbose",
    ]
    
    # Add permission mode: either --dangerously-skip-permissions or --permission-mode acceptEdits
    if dangerously_skip_permissions:
        cmd.append("--dangerously-skip-permissions")
    else:
        cmd.extend(["--permission-mode", "acceptEdits"])

    if model:
        cmd.extend(["--model", model])

    # Add --resume flag if session_id is provided
    if session_id is not None:
        cmd.extend(["--resume", session_id])

    # Unconditionally disallow orchestrator-level tools from managed agents
    cmd.extend(["--disallowed-tools", ",".join(DISALLOWED_TOOLS)])

    cmd.extend(["--", prompt])

    # Add any additional kwargs as command-line arguments
    for key, value in kwargs.items():
        if key == "timeout":
            # timeout is handled separately, not passed to CLI
            continue
        if key == "fork":
            # fork is converted to --fork-session flag
            if value is True:
                cmd.append("--fork-session")
            continue
        if value is True:
            cmd.append(f"--{key.replace('_', '-')}")
        elif value is not False and value is not None:
            cmd.append(f"--{key.replace('_', '-')}")
            cmd.append(str(value))
    
    return cmd


async def wrap_claude_code(
    prompt: str,
    model: Optional[str] = None,
    session_id: Optional[str] = None,
    timeout: Optional[float] = None,
    dangerously_skip_permissions: bool = False,
    cwd: Optional[str] = None,
    **kwargs
) -> Tuple[List[Dict[str, Any]], Optional[str]]:
    """
    Wraps claude code invocation in headless mode with stream-json output.
    This is an async function that can be run concurrently with other instances.
    
    NOTE: For production use, prefer wrap_claude_code_stream() which:
    - Enables progressive output processing (debug files, console output)
    - Uses idle timeout (resets on each data received) instead of total timeout
    - Allows detection of "stuck" executions while supporting long-running tasks
    
    This non-streaming version is retained for simpler test cases and backwards
    compatibility, but uses a TOTAL timeout rather than idle timeout.

    Args:
        prompt: The prompt to send to claude
        model: The model to use (e.g., "haiku", "sonnet", "opus")
        session_id: Optional session ID to resume an existing session (passes --resume flag)
        timeout: Optional total timeout in seconds (default: 600). Set to 0 for no timeout.
            WARNING: This is a TOTAL timeout, not an idle timeout. For long-running
            tasks that may exceed this duration, use wrap_claude_code_stream() instead.
        dangerously_skip_permissions: If True, passes --dangerously-skip-permissions
            instead of --permission-mode acceptEdits. WARNING: This allows Claude
            to execute any action without prompting for permission.
        cwd: Optional working directory for the subprocess. If None, inherits the
            orchestrator's working directory.
        **kwargs: Additional arguments to pass to claude command.
            Supported kwargs include:
            - fork (bool): If True, passes --fork-session flag to branch from session_id

    Returns:
        Tuple of (list of parsed JSON objects from the stream, extracted session_id or None)
        
    Raises:
        ClaudeCodeTimeoutError: If the total timeout is exceeded
        RuntimeError: If the command fails with non-zero exit code
    """
    # Build the command
    cmd = _build_claude_command(prompt, model, session_id, dangerously_skip_permissions, **kwargs)

    # Use default timeout if not specified
    # timeout=0 means no timeout (None for asyncio.wait_for)
    # timeout=None means use default
    if timeout == 0:
        effective_timeout = None
    elif timeout is not None:
        effective_timeout = timeout
    else:
        effective_timeout = DEFAULT_TIMEOUT

    # Run the command asynchronously and capture stdout
    # stdin=DEVNULL prevents the child from inheriting the terminal's stdin,
    # which would allow it to put the terminal in raw mode and disable SIGINT
    # generation from CTRL-C.
    process = await asyncio.create_subprocess_exec(
        *cmd,
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env=_build_claude_env(),
        cwd=cwd,
    )

    results = []
    extracted_session_id = None

    async def read_output():
        nonlocal extracted_session_id
        # Read in chunks to handle arbitrarily long lines
        # (default asyncio readline has a 64KB limit that can be exceeded by Claude Code)
        buffer = b""
        chunk_size = 1024 * 1024  # 1MB chunks
        
        while True:
            chunk = await process.stdout.read(chunk_size)
            if not chunk:
                # Process any remaining data in buffer
                if buffer:
                    line = buffer.decode('utf-8').strip()
                    if line:
                        try:
                            parsed = json.loads(line)
                            results.append(parsed)
                            if isinstance(parsed, dict):
                                if "session_id" in parsed:
                                    extracted_session_id = parsed["session_id"]
                                elif "metadata" in parsed and isinstance(parsed["metadata"], dict):
                                    if "session_id" in parsed["metadata"]:
                                        extracted_session_id = parsed["metadata"]["session_id"]
                        except json.JSONDecodeError as e:
                            logger.warning(f"Failed to parse JSON line: {line}", exc_info=e)
                break
            
            buffer += chunk
            
            # Process complete lines from buffer
            while b'\n' in buffer:
                line_bytes, buffer = buffer.split(b'\n', 1)
                line = line_bytes.decode('utf-8').strip()
                if not line:
                    continue

                try:
                    parsed = json.loads(line)
                    results.append(parsed)
                    
                    # Extract session_id from JSON objects if present
                    # Claude Code may output session_id in various formats
                    if isinstance(parsed, dict):
                        if "session_id" in parsed:
                            extracted_session_id = parsed["session_id"]
                        # Also check for nested session_id (e.g., in metadata)
                        elif "metadata" in parsed and isinstance(parsed["metadata"], dict):
                            if "session_id" in parsed["metadata"]:
                                extracted_session_id = parsed["metadata"]["session_id"]
                except json.JSONDecodeError as e:
                    # If a line isn't valid JSON, log it but continue
                    logger.warning(f"Failed to parse JSON line: {line}", exc_info=e)

    try:
        # Wait for output reading with timeout
        await asyncio.wait_for(read_output(), timeout=effective_timeout)
        
        # Wait for process to complete
        returncode = await asyncio.wait_for(
            process.wait(), 
            timeout=30  # Short timeout for cleanup after output is done
        )
    except asyncio.TimeoutError:
        logger.error(f"Claude Code timed out after {effective_timeout}s, killing process")
        process.kill()
        await process.wait()
        raise ClaudeCodeTimeoutError(
            f"Claude Code invocation timed out after {effective_timeout} seconds"
        )
    except BaseException:
        # Any other exception (CancelledError, KeyboardInterrupt, etc.)
        if process.returncode is None:
            process.kill()
            await process.wait()
        raise
    finally:
        # Last-resort cleanup: if the process is somehow still alive
        if process.returncode is None:
            try:
                process.kill()
                await process.wait()
            except Exception as e:
                logger.debug(f"Error during process cleanup: {e}")

    # Check for errors
    if returncode != 0:
        stderr_output = await process.stderr.read()
        stderr_text = stderr_output.decode('utf-8') if stderr_output else ""
        raise RuntimeError(
            f"Claude command failed with return code {returncode}\n"
            f"Stderr: {stderr_text}"
        )

    return results, extracted_session_id


async def wrap_claude_code_stream(
    prompt: str,
    model: Optional[str] = None,
    session_id: Optional[str] = None,
    timeout: Optional[float] = None,
    dangerously_skip_permissions: bool = False,
    cwd: Optional[str] = None,
    **kwargs
) -> AsyncIterator[Dict[str, Any]]:
    """
    Wraps claude code invocation and yields JSON objects as they arrive.
    This is an async generator that can be run concurrently with other instances.
    
    This is the PREFERRED way to invoke Claude Code as it:
    - Enables progressive output processing (debug files, console output)
    - Uses idle timeout (resets on each data received) instead of total timeout
    - Allows detection of "stuck" executions while supporting long-running tasks

    Args:
        prompt: The prompt to send to claude
        model: The model to use (e.g., "haiku", "sonnet", "opus")
        session_id: Optional session ID to resume an existing session (passes --resume flag)
        timeout: Optional idle timeout in seconds (default: 600). This is the maximum
            time to wait between receiving data chunks. Set to 0 for no timeout.
            Note: This is an IDLE timeout, not a total timeout. Long-running Claude
            Code executions will not timeout as long as they continue producing output.
        dangerously_skip_permissions: If True, passes --dangerously-skip-permissions
            instead of --permission-mode acceptEdits. WARNING: This allows Claude
            to execute any action without prompting for permission.
        cwd: Optional working directory for the subprocess. If None, inherits the
            orchestrator's working directory.
        **kwargs: Additional arguments to pass to claude command.
            Supported kwargs include:
            - fork (bool): If True, passes --fork-session flag to branch from session_id

    Yields:
        Parsed JSON objects from the stream as they arrive
        
    Raises:
        ClaudeCodeTimeoutError: If idle timeout is exceeded (no data received for timeout seconds)
        RuntimeError: If the command fails with non-zero exit code
    """
    # Build the command
    cmd = _build_claude_command(prompt, model, session_id, dangerously_skip_permissions, **kwargs)

    # Use default timeout if not specified
    # timeout=0 means no timeout (None)
    # timeout=None means use default
    if timeout == 0:
        idle_timeout = None
    elif timeout is not None:
        idle_timeout = timeout
    else:
        idle_timeout = DEFAULT_TIMEOUT

    # Run the command asynchronously and capture stdout
    # stdin=DEVNULL prevents the child from inheriting the terminal's stdin,
    # which would allow it to put the terminal in raw mode and disable SIGINT
    # generation from CTRL-C.
    process = await asyncio.create_subprocess_exec(
        *cmd,
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env=_build_claude_env(),
        cwd=cwd,
    )

    loop = asyncio.get_running_loop()
    last_data_time = loop.time()  # Track time of last data received for idle timeout

    # Read in chunks to handle arbitrarily long lines
    # (default asyncio readline has a 64KB limit that can be exceeded by Claude Code)
    buffer = b""
    chunk_size = 1024 * 1024  # 1MB chunks
    returncode = None
    
    try:
        while True:
            # Calculate remaining idle time for this read
            if idle_timeout is not None:
                elapsed_idle = loop.time() - last_data_time
                remaining_idle = idle_timeout - elapsed_idle
                
                if remaining_idle <= 0:
                    logger.error(
                        f"Claude Code idle timeout: no data received for {idle_timeout}s, killing process"
                    )
                    process.kill()
                    await process.wait()
                    raise ClaudeCodeTimeoutError(
                        f"Claude Code idle timeout: no data received for {idle_timeout} seconds"
                    )
                
                # Use remaining idle time as read timeout, capped at a reasonable max
                # to allow periodic idle timeout checks
                read_timeout = min(remaining_idle, 60.0)
            else:
                read_timeout = None  # No timeout
            
            try:
                if read_timeout is not None:
                    chunk = await asyncio.wait_for(
                        process.stdout.read(chunk_size),
                        timeout=read_timeout
                    )
                else:
                    chunk = await process.stdout.read(chunk_size)
            except asyncio.TimeoutError:
                # Read timed out - loop will check idle timeout on next iteration
                continue
            
            if not chunk:
                # EOF - process any remaining data in buffer
                if buffer:
                    line = buffer.decode('utf-8').strip()
                    if line:
                        try:
                            parsed = json.loads(line)
                            yield parsed
                        except json.JSONDecodeError as e:
                            logger.warning(f"Failed to parse final JSON line: {line}", exc_info=e)
                break
            
            # Data received - reset idle timer
            last_data_time = loop.time()
            buffer += chunk
            
            # Process complete lines from buffer
            while b'\n' in buffer:
                line_bytes, buffer = buffer.split(b'\n', 1)
                line = line_bytes.decode('utf-8').strip()
                if not line:
                    continue

                try:
                    parsed = json.loads(line)
                    yield parsed
                except json.JSONDecodeError as e:
                    # If a line isn't valid JSON, log it but continue
                    logger.warning(f"Failed to parse JSON line: {line}", exc_info=e)

        # Wait for process to complete (only reached if loop exits normally)
        returncode = await asyncio.wait_for(process.wait(), timeout=30)
    except GeneratorExit:
        # Generator is being closed (e.g., due to cancellation or CTRL-C)
        if process.returncode is None:
            logger.debug("Cleaning up Claude Code process due to generator close")
            try:
                process.kill()
                await process.wait()
            except Exception as e:
                logger.debug(f"Error during process cleanup on generator close: {e}")
        raise
    except asyncio.TimeoutError:
        logger.error(f"Claude Code cleanup timed out, killing process")
        process.kill()
        await process.wait()
        raise ClaudeCodeTimeoutError("Claude Code process cleanup timed out")
    finally:
        # Ensure process is cleaned up even if exception occurs during iteration
        if process.returncode is None:
            try:
                process.kill()
                await process.wait()
            except Exception as e:
                # Log but don't raise - we're in cleanup
                logger.debug(f"Error during process cleanup: {e}")
    
    # Check for errors (only if we got a returncode from normal completion)
    if returncode is not None and returncode != 0:
        stderr_output = await process.stderr.read()
        stderr_text = stderr_output.decode('utf-8') if stderr_output else ""
        raise RuntimeError(
            f"Claude command failed with return code {returncode}\n"
            f"Stderr: {stderr_text}"
        )

File: internal/ccwrap/ccwrap.go
// Package ccwrap wraps the claude CLI for use as a managed subprocess.
//
// Architecture:
//
//   - BuildClaudeCommand constructs the full argument list (testable without
//     running claude).
//   - BuildClaudeEnv copies the process environment and strips CLAUDECODE so
//     that claude does not treat itself as nested inside another session.
//   - InvokeStream is the preferred interface: it launches claude and returns a
//     channel of parsed JSON objects with an idle-timeout that resets on each
//     chunk received.
//   - Invoke is a convenience wrapper that collects all objects and extracts the
//     session ID; it uses a total timeout via context cancellation.
//
// The subprocess is started without a controlling terminal (setsid on Unix) so
// that claude's Node.js/Ink TUI cannot open /dev/tty and emit terminal control
// sequences that would interfere with raymond's console output.
package ccwrap

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
	"time"
)

const (
	// DefaultTimeout is the default idle/total timeout for Claude Code
	// invocations (10 minutes).
	DefaultTimeout = 600.0
)

// DisallowedTools lists tools that raymond-managed agents must never use;
// they are orchestrator-level concerns and must not be delegated to agents.
var DisallowedTools = []string{
	"EnterPlanMode", "ExitPlanMode", "AskUserQuestion", "NotebookEdit",
}

// claudeExe is the claude CLI executable name or path.
// Override in tests via overrideClaudeExe.
var claudeExe = "claude"

// ClaudeCodeTimeoutError is returned when a Claude Code invocation times out.
type ClaudeCodeTimeoutError struct {
	Timeout float64
	Idle    bool // true = idle timeout; false = total timeout
}

func (e *ClaudeCodeTimeoutError) Error() string {
	if e.Idle {
		return fmt.Sprintf(
			"Claude Code idle timeout: no data received for %.6g seconds", e.Timeout)
	}
	return fmt.Sprintf(
		"Claude Code invocation timed out after %.6g seconds", e.Timeout)
}

// BuildClaudeEnv returns a copy of the current process environment with the
// CLAUDECODE key removed so that claude does not treat itself as nested inside
// another session (raymond intentionally launches claude as a managed
// subprocess, not as a nested session).
func BuildClaudeEnv() map[string]string {
	env := make(map[string]string)
	for _, kv := range os.Environ() {
		if idx := strings.IndexByte(kv, '='); idx >= 0 {
			k := kv[:idx]
			if k == "CLAUDECODE" {
				continue
			}
			env[k] = kv[idx+1:]
		}
	}
	return env
}

// BuildClaudeCommand constructs the full claude CLI argument slice.
//
//   - model == ""     → omit --model
//   - effort == ""    → omit --effort
//   - sessionID == "" → omit --resume
//   - fork == true    → append --fork-session after the prompt separator
func BuildClaudeCommand(
	prompt, model, effort, sessionID string,
	dangerouslySkipPermissions, fork bool,
) []string {
	cmd := []string{
		claudeExe,
		"-p", // headless/print mode
		"--output-format", "stream-json",
		"--verbose",
	}

	if dangerouslySkipPermissions {
		cmd = append(cmd, "--dangerously-skip-permissions")
	} else {
		cmd = append(cmd, "--permission-mode", "acceptEdits")
	}

	if model != "" {
		cmd = append(cmd, "--model", model)
	}

	if effort != "" {
		cmd = append(cmd, "--effort", effort)
	}

	if sessionID != "" {
		cmd = append(cmd, "--resume", sessionID)
	}

	// Unconditionally prevent orchestrator-level tools from being used by
	// managed agents.
	cmd = append(cmd, "--disallowed-tools", strings.Join(DisallowedTools, ","))

	cmd = append(cmd, "--", prompt)

	if fork {
		cmd = append(cmd, "--fork-session")
	}

	return cmd
}

// ExtractSessionID returns the session ID embedded in a parsed JSON object.
// It checks the top-level "session_id" key first, then "metadata.session_id".
// Returns "" if no session ID is found.
func ExtractSessionID(obj map[string]any) string {
	if sid, ok := obj["session_id"].(string); ok && sid != "" {
		return sid
	}
	if meta, ok := obj["metadata"].(map[string]any); ok {
		if sid, ok := meta["session_id"].(string); ok && sid != "" {
			return sid
		}
	}
	return ""
}

// StreamItem holds a single parsed JSON object from the Claude Code stream,
// or an error if parsing or process execution failed.
type StreamItem struct {
	Object map[string]any
	Err    error
}

// InvokeStream launches a Claude Code subprocess and returns a channel that
// yields parsed JSON objects (StreamItem.Object) as they arrive on stdout.
//
//   - idleTimeout <= 0: no idle timeout
//   - idleTimeout > 0: sends ClaudeCodeTimeoutError if no data arrives for
//     that many seconds; the timer resets each time data is received
//   - cwd == "":        child inherits the caller's working directory
//
// The channel is always closed when the subprocess exits (or on error). The
// caller must drain the channel (or cancel ctx) to avoid goroutine leaks. A
// non-nil StreamItem.Err in the last item indicates subprocess failure or
// timeout.
func InvokeStream(
	ctx context.Context,
	prompt, model, effort, sessionID string,
	idleTimeout float64,
	dangerouslySkipPermissions, fork bool,
	cwd string,
) <-chan StreamItem {
	ch := make(chan StreamItem, 64)
	go func() {
		defer close(ch)
		err := runStream(ctx, ch, prompt, model, effort, sessionID, idleTimeout,
			dangerouslySkipPermissions, fork, cwd)
		if err != nil {
			select {
			case ch <- StreamItem{Err: err}:
			case <-ctx.Done():
				// Caller already cancelled; don't block sending the error.
			}
		}
	}()
	return ch
}

// lineMsg is an internal message from the scanner goroutine.
type lineMsg struct {
	line string
	err  error
}

func runStream(
	ctx context.Context,
	ch chan<- StreamItem,
	prompt, model, effort, sessionID string,
	idleTimeout float64,
	dangerouslySkipPermissions, fork bool,
	cwd string,
) error {
	cmdSlice := BuildClaudeCommand(prompt, model, effort, sessionID, dangerouslySkipPermissions, fork)
	execCmd := exec.Command(cmdSlice[0], cmdSlice[1:]...)
	execCmd.Stdin = nil // → /dev/null per exec.Cmd docs

	envMap := BuildClaudeEnv()
	envSlice := make([]string, 0, len(envMap))
	for k, v := range envMap {
		envSlice = append(envSlice, k+"="+v)
	}
	execCmd.Env = envSlice

	if cwd != "" {
		execCmd.Dir = cwd
	}

	// Platform-specific: detach from controlling terminal.
	setupClaudeCmd(execCmd)

	stdoutPipe, err := execCmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("stdout pipe: %w", err)
	}
	var stderrBuf bytes.Buffer
	execCmd.Stderr = &stderrBuf

	if err := execCmd.Start(); err != nil {
		return fmt.Errorf("failed to start claude: %w", err)
	}

	// Scanner goroutine — reads stdout line by line and sends to lineCh.
	// It closes lineCh when done, signalling the main loop to call Wait().
	lineCh := make(chan lineMsg, 32)
	go func() {
		defer close(lineCh)
		scanner := bufio.NewScanner(stdoutPipe)
		scanner.Buffer(make([]byte, 1024*1024), 1024*1024) // 1 MB max line length
		for scanner.Scan() {
			lineCh <- lineMsg{line: scanner.Text()}
		}
		if err := scanner.Err(); err != nil {
			lineCh <- lineMsg{err: err}
		}
	}()

	// Idle timeout setup.
	var idleTimer *time.Timer
	var idleExpired <-chan time.Time
	resetIdle := func() {}
	if idleTimeout > 0 {
		d := time.Duration(float64(time.Second) * idleTimeout)
		idleTimer = time.NewTimer(d)
		idleExpired = idleTimer.C
		resetIdle = func() {
			if !idleTimer.Stop() {
				select {
				case <-idleTimer.C:
				default:
				}
			}
			idleTimer.Reset(d)
		}
		defer idleTimer.Stop()
	}

	// killAndDrain kills the entire process group (so orphaned child processes
	// cannot hold the stdout/stderr pipes open), waits for process exit, then
	// drains lineCh so the scanner goroutine can exit cleanly.
	killAndDrain := func() {
		_ = killClaudeProcess(execCmd)
		_ = execCmd.Wait()
		for range lineCh {}
	}

	for {
		select {
		case <-ctx.Done():
			killAndDrain()
			return ctx.Err()

		case <-idleExpired:
			// Prefer context cancellation over idle timeout when both fire
			// simultaneously (e.g. total timeout deadline expired at the same
			// moment the idle timer fired).
			if ctx.Err() != nil {
				killAndDrain()
				return ctx.Err()
			}
			killAndDrain()
			return &ClaudeCodeTimeoutError{Timeout: idleTimeout, Idle: true}

		case msg, ok := <-lineCh:
			if !ok {
				// Scanner goroutine closed lineCh → stdout EOF; process exited.
				if waitErr := execCmd.Wait(); waitErr != nil {
					if exitErr, ok2 := waitErr.(*exec.ExitError); ok2 {
						return fmt.Errorf(
							"claude command failed with return code %d\nStderr: %s",
							exitErr.ExitCode(), stderrBuf.String())
					}
					if ctx.Err() != nil {
						return ctx.Err()
					}
					return fmt.Errorf("claude Wait: %w", waitErr)
				}
				return nil
			}

			if msg.err != nil {
				killAndDrain()
				return fmt.Errorf("reading claude output: %w", msg.err)
			}

			resetIdle()

			line := strings.TrimSpace(msg.line)
			if line == "" {
				continue
			}
			var obj map[string]any
			if err := json.Unmarshal([]byte(line), &obj); err != nil {
				// Skip non-JSON lines (mirrors Python's warning + continue).
				log.Printf("ccwrap: skipping non-JSON line from claude: %q", line)
				continue
			}

			select {
			case ch <- StreamItem{Object: obj}:
			case <-ctx.Done():
				killAndDrain()
				return ctx.Err()
			}
		}
	}
}

// Invoke runs Claude Code non-streamingly, collecting all JSON objects and
// returning them along with the extracted session ID.
//
//   - totalTimeout <= 0: no timeout (relies on any deadline in ctx)
//   - totalTimeout > 0: ClaudeCodeTimeoutError if total duration exceeded
func Invoke(
	ctx context.Context,
	prompt, model, effort, sessionID string,
	totalTimeout float64,
	dangerouslySkipPermissions, fork bool,
	cwd string,
) ([]map[string]any, string, error) {
	if totalTimeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(
			ctx, time.Duration(float64(time.Second)*totalTimeout))
		defer cancel()
	}

	ch := InvokeStream(ctx, prompt, model, effort, sessionID, 0,
		dangerouslySkipPermissions, fork, cwd)

	var results []map[string]any
	var extractedSID string
	for item := range ch {
		if item.Err != nil {
			if errors.Is(item.Err, context.DeadlineExceeded) {
				return nil, "", &ClaudeCodeTimeoutError{Timeout: totalTimeout, Idle: false}
			}
			return nil, "", item.Err
		}
		results = append(results, item.Object)
		if sid := ExtractSessionID(item.Object); sid != "" {
			extractedSID = sid
		}
	}

	// Safety check: handle timeout if the error was dropped by InvokeStream's
	// select (both ctx.Done() and ch send were simultaneously ready).
	if totalTimeout > 0 && errors.Is(ctx.Err(), context.DeadlineExceeded) {
		return nil, "", &ClaudeCodeTimeoutError{Timeout: totalTimeout, Idle: false}
	}

	return results, extractedSID, nil
}
